### ISSUE TITLE
Pytorch Implementation

### ISSUE THREAD

**Issue by luke-mcdermott-mi** [yolky/RCIG#4]
Do you plan to release a pytorch implementation of this code?

I am trying to replicate this as closely as possible in pytorch, using the same code for everything and making my own replicas; however, this only gets ~27% accuracy with conv. For loss functions, are you just using MSE? I see the l2 code; however, you just set l2 to 0. I am not finding where this makes a difference. 

All I actually need is just some train(my_model) and a working pytorch test_loader that evaluates this, but there is a lot of bloat in this code.

Here is my implementation. I have no clue what could be causing this. I use your data, so ZCA should already be done. I use adam optimizer instead; however, this should drop the accuracy by that much. Everything outside of this code is your code in jax.

```
 class CoresetDataset(Dataset):
    def __init__(self, images, labels):
        self.images = images
        self.labels = labels

    def __len__(self):
        return len(self.images)

    def __getitem__(self, idx):
        image = self.images[idx].transpose((2, 0, 1))
        label = self.labels[idx]
        return torch.tensor(image, dtype=torch.float32), torch.tensor(self.labels[idx], dtype=torch.float32)



class TensorFlowToTorchDataset(Dataset):
    def __init__(self, tf_dataset):
        images = []
        labels = []
        
        for image, label in tf_dataset.as_numpy_iterator(): # <- Add this line
            images.append(image)
            labels.append(label)

        self.images = np.concatenate(images, axis=0)
        self.labels = np.concatenate(labels, axis=0)

    def __len__(self):
        return len(self.images)

    def __getitem__(self, idx):
        image, label = self.images[idx], self.labels[idx]
        return torch.tensor(image.transpose((2, 0, 1)), dtype=torch.float32), torch.tensor(self.labels[idx], dtype=torch.float32)


def torch_eval(dataset_name='cifar10', data_path=None, zca_path=None, train_log=None, train_img=None, width=128, depth=3, normalization='identity', eval_lr=0.0001, random_seed=0, message='eval_log', output_dir=None, max_cycles=1000, config_path=None, checkpoint_path=None, save_name='eval_result', log_dir=None, eval_arch='resnet', models_to_test=5):
    # --------------------------------------
    # Setup
    # --------------------------------------
    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") 

    if output_dir is None:
        output_dir = os.path.dirname(checkpoint_path)

    if log_dir is None:
        log_dir = output_dir

    logging.use_absl_handler()

    logging.get_absl_handler().use_absl_log_file('{}, {}'.format(int(time.time()), message), './{}/'.format(log_dir))
    absl.flags.FLAGS.mark_as_parsed() 
    logging.set_verbosity('info')
    
    logging.info('\n\n\n{}\n\n\n'.format(message))
    
    config = get_config()
    #config.kernel.batch_size = 256
    config.random_seed = random_seed
    config.train_log = train_log if train_log else 'train_log'
    config.train_img = train_img if train_img else 'train_img'


    config.dataset.data_path = data_path if data_path else 'data/tensorflow_datasets'
    config.dataset.zca_path = zca_path if zca_path else 'data/zca'
    config.dataset.name = dataset_name

    (ds_train, ds_test), preprocess_op, rev_preprocess_op, proto_scale = get_dataset(config.dataset)
    
    y_transform = lambda y: tf.one_hot(y, config.dataset.num_classes, on_value=1 - 1 / config.dataset.num_classes,
                                    off_value=-1 / config.dataset.num_classes)
    ds_train = configure_dataloader(ds_train, batch_size=config.kernel.batch_size, y_transform=y_transform,
                                        train=True, shuffle=True)
    ds_test = configure_dataloader(ds_test, batch_size=config.kernel.eval_batch_size, y_transform=y_transform,
                                   train=False, shuffle=False)
    
    num_classes = config.dataset.num_classes

    if config.dataset.img_shape[0] in [28, 32]:
        depth = 3
    elif config.dataset.img_shape[0] == 64:
        depth = 4
    elif config.dataset.img_shape[0] == 128:
        depth = 5
    else:
        raise Exception('Invalid resolution for the dataset')


    loaded_checkpoint = checkpoints.restore_checkpoint(f'./{checkpoint_path}', None)
    coreset_images = loaded_checkpoint['ema_average']['x_proto']
    coreset_labels = loaded_checkpoint['ema_average']['y_proto']
    coreset_dataset = CoresetDataset(coreset_images, coreset_labels)
    train_loader = DataLoader(coreset_dataset, batch_size=config.kernel.batch_size, shuffle=True)
    
    # Get the length of the TensorFlow dataset
    # Pass the whole TensorFlow dataset
    torch_ds_test = TensorFlowToTorchDataset(ds_test)
    # Pass the PyTorch dataset to the DataLoader
    test_loader = DataLoader(torch_ds_test, batch_size=config.kernel.eval_batch_size, shuffle=False)

    
    if eval_arch == 'conv':
        model = ConvNet(3, 10, 64, 3, 'relu', 'batchnorm', 'maxpooling').to(device)
        
        #Conv(use_softplus = False, beta = 20., num_classes = num_classes, width = width, depth = depth, normalization = normalization)
    elif eval_arch == 'resnet':
        model = resnet18()
        model.fc = nn.Linear(512,10)
        model = model.to(device)

    # Replace the optimizer initialization with PyTorch's optimizer
    optimizer = Adam(model.parameters(), lr=eval_lr)

    # Replace the learning rate scheduler with PyTorch's scheduler
    num_online_eval_updates = 1000 if coreset_images.shape[0] == 10 else 2000
    lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=num_online_eval_updates, eta_min=0.01)
    criterion = nn.MSELoss() #torch.nn.BCEWithLogitsLoss()

    # Training loop should be modified to work with PyTorch
    for epoch in range(max_cycles):
        # Training
        model.train()
        for batch_idx, (images, labels) in enumerate(train_loader):
            images = images.to(device)
            labels = labels.to(device)
            outputs = model(images)
            #loss, _ = torch_get_training_loss_l2(model, images, labels, l2=0.0)
            loss = criterion(outputs, labels)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

        # Update the learning rate
        lr_scheduler.step()
        torch.cuda.empty_cache()

    # Evaluation
    model.eval()
    with torch.no_grad():
        correct = 0
        total = 0
        for i, (images, labels) in enumerate(test_loader): 
            images, labels = images.to(device), labels.to(device)
            test_output = model(images)

            
            pred_y = torch.max(test_output, 1)[1].data.squeeze()
            true_y = labels

            
            correct += (pred_y == true_y.argmax(dim=1)).sum().item()
            total += labels.size(0)
        accuracy = correct / total
        print('Accuracy: ', accuracy)

    return model
```

and this is my conv class

```
class ConvNet(nn.Module):
    def __init__(self, channel, num_classes, net_width, net_depth, net_act, net_norm, net_pooling, im_size = (32,32)):
        super(ConvNet, self).__init__()

        self.features, shape_feat = self._make_layers(channel, net_width, net_depth, net_norm, net_act, net_pooling, im_size)
        num_feat = shape_feat[0]*shape_feat[1]*shape_feat[2]
        self.classifier = nn.Linear(num_feat, num_classes)

    def forward(self, x):
        # print("MODEL DATA ON: ", x.get_device(), "MODEL PARAMS ON: ", self.classifier.weight.data.get_device())
        out = self.features(x)
        out = out.reshape(out.size(0), -1) #was view
        out = self.classifier(out)
        return out

    def _get_activation(self, net_act):
        if net_act == 'sigmoid':
            return nn.Sigmoid()
        elif net_act == 'relu':
            return nn.ReLU(inplace=True)
        elif net_act == 'leakyrelu':
            return nn.LeakyReLU(negative_slope=0.01)
        else:
            exit('unknown activation function: %s'%net_act)

    def _get_pooling(self, net_pooling):
        if net_pooling == 'maxpooling':
            return nn.MaxPool2d(kernel_size=2, stride=2)
        elif net_pooling == 'avgpooling':
            return nn.AvgPool2d(kernel_size=2, stride=2)
        elif net_pooling == 'none':
            return None
        else:
            exit('unknown net_pooling: %s'%net_pooling)

    def _get_normlayer(self, net_norm, shape_feat):
        # shape_feat = (c*h*w)
        if net_norm == 'batchnorm':
            return nn.BatchNorm2d(shape_feat[0], affine=True)
        elif net_norm == 'layernorm':
            return nn.LayerNorm(shape_feat, elementwise_affine=True)
        elif net_norm == 'instancenorm':
            return nn.GroupNorm(shape_feat[0], shape_feat[0], affine=True)
        elif net_norm == 'groupnorm':
            return nn.GroupNorm(4, shape_feat[0], affine=True)
        elif net_norm == 'none':
            return None
        else:
            exit('unknown net_norm: %s'%net_norm)

    def _make_layers(self, channel, net_width, net_depth, net_norm, net_act, net_pooling, im_size):
        layers = []
        in_channels = channel
        if im_size[0] == 28:
            im_size = (32, 32)
        shape_feat = [in_channels, im_size[0], im_size[1]]
        for d in range(net_depth):
            layers += [nn.Conv2d(in_channels, net_width, kernel_size=3, padding=3 if channel == 1 and d == 0 else 1)]
            shape_feat[0] = net_width
            if net_norm != 'none':
                layers += [self._get_normlayer(net_norm, shape_feat)]
            layers += [self._get_activation(net_act)]
            in_channels = net_width
            if net_pooling != 'none':
                layers += [self._get_pooling(net_pooling)]
                shape_feat[1] //= 2
                shape_feat[2] //= 2


        return nn.Sequential(*layers), shape_feat
```
